### ISSUE TITLE
update refactoring provider to have something that we can use to order its listing in lightbulb

### ISSUE THREAD

**Issue by heejaechang** [dotnet/roslyn#33818]
currently, we distinguish refactoring and code fix but in reality, it doesn't matter much for users.

code fix can use diagnostic and caret location to figure out what's more important and then use priority to sort things in light bulb menu.

but refactoring doesn't have such mark (anchor) like diagnostic for code fix that we can use to place it in right place in menu.

so we do something like this - http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/Suggestions/SuggestedActionsSource.cs,197

and distinguish refactoring and code fix. I believe this will be just a source of an endless tweak as new refactoring is added and we want to place it in a different place. 

I think we should fix the root cause not just keep changing ordering rules. also, we don't actually have tests for all different combinations of selection and refactorings and code fix mix so we actually have no idea how/what behavior got changed due to this kind of tweak.

we should discuss how we should address this once for all in a design meeting.

...

my suggestion will be making code refactoring and code fix to behave the same (this theme is consistent with this - https://github.com/dotnet/roslyn/issues/32461) and order those without distinguishing two.

right now, the only difference between these 2 for ordering is the applicable span for suggestedactionset.

code fix has the applicable span from diagnostic it is fixing, but refactoring gets it from selection (which might or might not directly related to the refactoring and it get even wierdier if there is no explicit selection) 

due to that, ordering those 2 (code fix and code refactoring) without distinguishing them mess things up. and we had to do this hueristic based ordering we have even though we don't actually know whether selection is used for refactoring or not.

if we see code fix provider, code fix is registered this way (or variation of the API)
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/CodeFixContext.cs,165

```csharp
public void RegisterCodeFix(CodeAction action, Diagnostic diagnostic)
```

and refactoring is done this way
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeRefactorings/CodeRefactoringContext.cs,51

```csharp
public void RegisterRefactoring(CodeAction action)
```

I suggest we add an overload (or its variation) for coderefactoring provider

```csharp
public void RegisterRefactoring(CodeAction action, TextSpan span)
```

and we use that span for applicable span for suggestedActionSet and for those refactoring that added this way, we don't disthinguish it from code fixes when ordering.

also, another benefit of this is we no longer need to do the hard to discover "caret must be on this specific magic location for refactoring to show up in LB".

now we can include refactoring in menu if the span provided is on the line where caret is on, exact same behavior as code fix.

we can leave existing API as it is, so refactoring that truely depends on selection like extract method can still use it.  

...

the suggestion above is something I just think on top of my head so we can discuss more in design meeting.

**Comment by heejaechang** [comment]
tagging @dotnet/roslyn-ide @sharwell @mavasani @jinujoseph

**Comment by miloush** [comment]
I agree that it is frustrating having the order (especially the top item which gets into the muscle memory) changed every release.

I wonder if it might be worth allowing the user to influence the ordering or select which refactoring/fixes should be offered.

**Comment by CyrusNajmabadi** [comment]
> and distinguish refactoring and code fix. I believe this will be just a source of an endless tweak as new refactoring is added and we want to place it in a different place.

It hasn't been a source of endless tweaks.  We've added several dozens of features over the years, and in general, we've only needed to make tiny tweaks occasionally.

> my suggestion will be making code refactoring and code fix to behave the same (this theme is consistent with this - #32461) and order those without distinguishing two.

I disagree.  A core distinction we found was that people expected to be shown the fixes for the issues present on the line.  So, i think what we want is:

1. if there is a *visible* adornment on the line, then that should influence priority.
2. if there is a selection, that should influence things.  if i select some code, it's a large signal i want to do something with that code.  

So, it would be ok for a refactoring to be sorted across invisible-fixes.  But visible fixes should take priority as long as there isn't a selection.

--

Note: this ties into previous PRs i've tried to provide which unify the concept of "a refactoring" with "a codefix that doesn't have an adornment".

--

> code fix has the applicable span from diagnostic it is fixing, but refactoring gets it from selection (which might or might not directly related to the refactoring

I'm not sure i understand this.  Which refactorings are not directly relates to the selection/caret the user is at?  That was the whole idea of refactorings and hte refactoring API :)

--

> I suggest we add an overload (or its variation) for coderefactoring provider

What would this TextSpan be for?

**Comment by CyrusNajmabadi** [comment]
A core problem i have with this issue is that it doesn't actually demonstrate how it would resolve the issue that prompted things.  i.e.:

https://github.com/dotnet/roslyn/pull/33631#issuecomment-468902695

It also doesn't show how it would have resolved the other times where we manually tweaked things to address ordering.  Instead of discussing "the solution" first, can we provide examples of the times we've had this problem and *then* show how the solution will be sufficient?

**Comment by CyrusNajmabadi** [comment]
> the suggestion above is something I just think on top of my head so we can discuss more in design meeting.

can you please include me in such a meeting.  i'm happy to call in as necessary.  I have a lot of familiarity with this side of things, especially how much effort we've put in over the years *at many levels* to get good lists.

**Comment by CyrusNajmabadi** [comment]
Ii mentioned this in the original PR, but it's relevant here as well:

--

> I believe this kind of heuristic will require an endless tweak as new refactoring are added.

Note: this was not a tweak.  This was literally just a bug.  The problem here was that we have a system for prioritizing refactorings.  However, the implementation was buggy, and a low-pri refactoring wasn't showing up after medium pri refactorings *if* that refactoring had nested refactorings.  

Nothing about the actual prioritization design changed here.  We just fixed up the impl to match teh intended design.  So i question if anything needs to be done here.

**Comment by CyrusNajmabadi** [comment]
Closing out.  Through the work @petrroll did, we have a sytem that works quite well for providing a more consistent approach for refactorings.  We've also been able to tweak very easily as we've encountered undesirable behavior.  it's unclear what would actually benefit from here as we wwould need the new api, and we would still have to go to all refactorings to have them decide how to do it.  At that point, just fixing the refactorings with the existing system would be a viable path forward.
